x/orderbook/handler.go

```diff
@@ -105,3 +113,219 @@ func (h OrderBookHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.
+	return CreateOrderHandler{
+		auth:            auth,
+		mover:           mover,
+		orderBucket:     NewOrderBucket(),
+		orderBookBucket: NewOrderBookBucket(),
+		marketBucket:    NewMarketBucket(),
+		controller: controller{
+			orderBucket: NewOrderBucket(),
+			tradeBucket: NewTradeBucket(),
+			mover:       mover,
+		},
+	}
+}
+
+// Check just verifies it is properly formed and returns
+// the cost of executing it.
+func (h CreateOrderHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	_, _, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &weave.CheckResult{GasAllocated: newOrderCost}, nil
+}
+
+// validate does all common pre-processing between Check and Deliver
+func (h CreateOrderHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*CreateOrderMsg, *OrderBook, error) {
+	var msg CreateOrderMsg
+
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, nil, errors.Wrap(err, "load msg")
+	}
+
+	// Get rules from the orderbook
+	var orderBook OrderBook
+	if err := h.orderBookBucket.One(db, msg.OrderBookID, &orderBook); err != nil {
+		return nil, nil, errors.Wrap(err, "load orderbook")
+	}
+
+	// make sure we have a valid ticker
+	t := msg.Offer.Ticker
+	if t != orderBook.AskTicker && t != orderBook.BidTicker {
+		return nil, nil, errors.Wrap(errors.ErrCurrency, "offer ticker not in this orderbook")
+	}
+
+	// ensure they can make this trade
+	if msg.Trader != nil && !h.auth.HasAddress(ctx, msg.Trader) {
+		return nil, nil, errors.Wrap(errors.ErrUnauthorized, "must be authorized by the Trader")
+	}
+
+	// TODO: do we want to check the market it belongs to, to check fee payments, etc?
+
+	return &msg, &orderBook, nil
+}
+
+// Deliver creates an orderbook and saves if all preconditions are met
+func (h CreateOrderHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, ob, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	blockTime, err := weave.BlockTime(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "no block time in header")
+	}
+	now := weave.AsUnixTime(blockTime)
+
+	trader := msg.Trader
+	if trader == nil {
+		trader = x.MainSigner(ctx, h.auth).Address()
+	}
+
+	side := Side_Ask
+	if msg.Offer.Ticker == ob.BidTicker {
+		side = Side_Bid
+	}
+
+	order := &Order{
+		Metadata:       &weave.Metadata{},
+		OrderBookID:    msg.OrderBookID,
+		Trader:         trader,
+		Side:           side,
+		OrderState:     OrderState_Open,
+		OriginalOffer:  msg.Offer,
+		RemainingOffer: msg.Offer,
+		Price:          msg.Price,
+		CreatedAt:      now,
+		UpdatedAt:      now,
+	}
+
+	// Save Order
+	if err := h.orderBucket.Put(db, order); err != nil {
+		return nil, err
+	}
+
+	// Send money to contract
+	// We must calculate the address after saving to have proper auto-generated ID
+	err = h.mover.MoveCoins(db, trader, order.Address(), *msg.Offer)
```

Comment:
Interesting... Can you elaborate why we move the coins to `order.Address`? I guess trader's coins must be deposited to another account so he cant make changes on it while order is awaiting.

Reply:
Yes. We need to store them somewhere that is controlled by the logic of our extension.

I could make an orderbook account and move them all there.
However, I figured moving to an Order is safer. If there is some error in the logic, I can empty coins in the order, but no more. I cannot extract from other orders. (Assuming some specially crafted Order exposes a malicious bug, you cannot create one yourself and empty out other orders). 

Idea is to give the code least access that it needs - just the coins locked to this order.